File: JavaScript/99.恢复二叉搜索树.js
/*
 * @lc app=leetcode.cn id=99 lang=javascript
 *
 * [99] 恢复二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
/*
Accepted
1919/1919 cases passed (172 ms)
Your runtime beats 13.86 % of javascript submissions
Your memory usage beats 21.09 % of javascript submissions (47.5 MB)
*/
var recoverTree = function (root) {
  const nums = []
  const inorder = node => {
    if (!node) return
    inorder(node.left)
    nums.push(node.val)
    inorder(node.right)
  }
  inorder(root)

  const findTowSwapped = () => {
    let left = -1, right = -1
    for (let i = 0; i < nums.length; i++) {
      const num = nums[i]
      if (left < 0) {
        if (num > nums[i + 1]) left = i
      } else {
        if (num < nums[i - 1]) right = i
      }
    }
    return [nums[left], nums[right]]
  }

  const [left, right] = findTowSwapped()
  const recover = (node, count, x, y) => {
    if (!node) return
    if (node.val === x || node.val === y) {
      node.val = node.val === x ? y : x
      if (--count === 0) return
    }
    recover(node.left, count, x, y)
    recover(node.right, count, x, y)
  }
  recover(root, 2, left, right)
};
// @lc code=end

File: Python3/99.恢复二叉搜索树.py
#
# @lc app=leetcode.cn id=99 lang=python3
#
# [99] 恢复二叉搜索树
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

""" 
python的骚操作
Accepted
1919/1919 cases passed (52 ms)
Your runtime beats 83.47 % of python3 submissions
Your memory usage beats 32.58 % of python3 submissions (15.3 MB)
"""


class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        left, right = None, None
        prev = None

        def handler(node):
            nonlocal prev, left, right
            if not node:
                return
            handler(node.left)
            if not prev:
                prev = node
            else:
                if node.val < prev.val:
                    right, left = node, prev if left is None else left
                prev = node
            handler(node.right)

        handler(root)
        if left and right:
            left.val, right.val = right.val, left.val

# @lc code=end
